torch/_guards.py

```diff
@@ -52,17 +53,28 @@
 """
 
 
-class CompileId(NamedTuple):
-    frame_id: int
+# TODO: mark as kw_only=True once we drop support for <Python 3.10
+@dataclass(frozen=True)
+class CompileId:
+    frame_id: Optional[int]
     # This id is per-frame, and counts how many times we've compiled this
     # frame.  This could have been a global id but having this be per-frame
     # gives you a better intuitive sense for how many recompiles have occurred
     # so far.
-    frame_compile_id: int
+    frame_compile_id: Optional[int]
     # TODO: consider also tracking the recompilation count
+    compiled_autograd_id: Optional[int] = None
 
     def __str__(self):
-        return f"{self.frame_id}/{self.frame_compile_id}"
+        # Keep this in sync with tlparse repo
+        if self.compiled_autograd_id is not None:
+            assert (self.frame_id is None) == (self.frame_compile_id is None)
+            return f"!{self.compiled_autograd_id}/{self.frame_id}/{self.frame_compile_id}".replace(
+                "None", "-"
```

Comment:
Would probably be beneficial to drop self.frame_id/self.frame_compile_id entirely when they are both None

Reply:
it makes preserving uniqueness a bit more dangerous, newly added fields would either need a special prefix/suffix or not elide themselves

Reply:
But you are adding a special prefix so it's unambiguous right? 8)